src/integrationtests/java/com/aws/greengrass/integrationtests/lifecyclemanager/GenericExternalServiceIntegTest.java

```diff
@@ -620,6 +640,84 @@ void GIVEN_linux_resource_limits_WHEN_it_changes_THEN_component_runs_with_new_re
         assertResourceLimits(componentName, 10240l * 1024, 1.5);
     }
 
+    @EnabledOnOs({OS.LINUX})
+    @Test
+    void GIVEN_linux_resource_limits_WHEN_it_changes_THEN_component_runs_with_new_resource_limits_V2() throws Exception {
+        String echoComponentName = "echo_service";
+        String mainComponentName = "main";
+        String rootGGPathString = ROOT_PATH_STRING + "/" + GG_PATH_STRING;
+        String componentPathString = rootGGPathString + "/" + echoComponentName;
+        // Run with no resource limit
+        ConfigPlatformResolver.initKernelWithMultiPlatformConfig(kernel,
+                getClass().getResource("config_run_with_user.yaml"));
+        CountDownLatch service = new CountDownLatch(1);
+        kernel.getContext().addGlobalStateChangeListener((s, oldState, newState) -> {
+            if (s.getName().equals(echoComponentName) && newState.equals(State.RUNNING)) {
+                service.countDown();
+            }
+        });
+
+        linuxPlatform = spy(kernel.getContext().get(LinuxPlatform.class));
+
+        createComponentData(echoComponentName);
+        createComponentData(mainComponentName);
+
+        // Due to cgroup v1 is active by default (in test platform), and the directories of cgroup v1 are read-only
+        // therefore, here create some directories and files as fake cgroup v2 files to support testing
+        Field controllerField = LinuxPlatform.class.getDeclaredField("CGROUP_CONTROLLERS");
+        setFinalStatic(controllerField, Paths.get(componentPathString + "/memory.max"));
+        systemResourceController = linuxPlatform.getSystemResourceController();
+        LinuxSystemResourceControllerV2 controllerV2 = (LinuxSystemResourceControllerV2) systemResourceController;
+        Field memoryCgroupField = LinuxSystemResourceControllerV2.class.getSuperclass().getDeclaredField("memoryCgroup");
+        memoryCgroupField.setAccessible(true);
+        Cgroup memoryCgroup = (Cgroup) memoryCgroupField.get(controllerV2);
+        Field subsystem = memoryCgroup.getClass().getDeclaredField("subSystem");
+        subsystem.setAccessible(true);
+        CgroupSubSystemV2 cg = (CgroupSubSystemV2) subsystem.get(memoryCgroup);
+        Field f = cg.getClass().getInterfaces()[0].getDeclaredField("CGROUP_ROOT");
+        setFinalStatic(f, Paths.get(ROOT_PATH_STRING));
```

Comment:
why are you using reflection?

Even in testing, reflection should not be used.

Reply:
we are confused about how to do it without reflection. Since the automation test platform has Cgroup V1 being activated, thus we have to simulate Cgroup V2 is active. And because the folders of Cgroup V1 are read-only, therefore we have to change CGROUP_ROOT path to a test path so that the folder/file structures of V2 can be created under that test root path. But due to CGROUP_ROOT is a final field in interface “CGroupSubSystemPath.java”, so without reflection, this path is not able to be reset to the test path. We also tried mock/stub with @InjectionMocks, @Spy, @Mock, but as kernel.lanuch() is the core method called in integration test, it is too deep for stubbing to get effective. Could you please also have a look at this issue, if there is any other approach to change the CGROUP_ROOT path?

Reply:
What I'd want in this case is that the test is skipped on OSes which do not have Cgroup v2. When running on a device which does have cgroup v2 enabled, then I'd want the v2 test to run and not the v1 test.